### PR TITLE
Add biased topk

### PR DESCRIPTION
**Pull request by gaopengff**
Adapted from cuda implementation in sglang [moe_fused_gate.cuh](https://github.com/sgl-project/sglang/blob/v0.5.15/python/sglang/jit_kernel/csrc/moe/moe_fused_gate.cuh). The performance comparison between this kernel and default implementation(triton with torch.compile) in sglang are as follows.

Configs are from deepseek-v4:
|   num_tokens |   num_experts |   topk | dtype         | scoring_func   | renormalize   |   num_shared |   scale | apply_scale   |   sglang_kernel_ms |   triton_ms |   speedup |
|-------------:|--------------:|-------:|:--------------|:---------------|:--------------|-------------:|--------:|:--------------|-------------------:|------------:|----------:|
|            1 |           256 |      6 | torch.float32 | sqrtsoftplus   | True          |            1 |     2.5 | True          |           0.011752 |    0.041704 |      3.55 |
|            1 |           384 |      6 | torch.float32 | sqrtsoftplus   | True          |            1 |     2.5 | True          |           0.011024 |    0.039208 |      3.56 |
|           32 |           256 |      6 | torch.float32 | sqrtsoftplus   | True          |            1 |     2.5 | True          |           0.01118  |    0.042796 |      3.83 |
|           32 |           384 |      6 | torch.float32 | sqrtsoftplus   | True          |            1 |     2.5 | True          |           0.011336 |    0.043524 |      3.84 |
|          256 |           256 |      6 | torch.float32 | sqrtsoftplus   | True          |            1 |     2.5 | True          |           0.02366  |    0.05798  |      2.45 |
|          256 |           384 |      6 | torch.float32 | sqrtsoftplus   | True          |            1 |     2.5 | True          |           0.03016  |    0.06994  |      2.32 |
|         1024 |           256 |      6 | torch.float32 | sqrtsoftplus   | True          |            1 |     2.5 | True          |           0.015496 |    0.13819  |      8.92 |
|         1024 |           384 |      6 | torch.float32 | sqrtsoftplus   | True          |            1 |     2.5 | True          |           0.029588 |    0.172588 |      5.83 |
|         8192 |           256 |      6 | torch.float32 | sqrtsoftplus   | True          |            1 |     2.5 | True          |           0.084188 |    0.818012 |      9.72 |
|         8192 |           384 |      6 | torch.float32 | sqrtsoftplus   | True          |            1 |     2.5 | True          |           0.158444 |    1.10942  |      7    |

### CODE REVIEW

File: python/sgl_kernel/moe.py [hunk 1]
```diff
@@ -143,6 +144,37 @@ def moe_fused_gate(
...
+    num_fused_shared_experts=0,
+    renormalize=False,
+    routed_scaling_factor=1.0,
+    apply_routed_scaling_factor_on_output=False,
+):
+    scoring_func_int = _MOE_SCORING_FUNC_MAP.get(scoring_func.lower())
+    assert (
+        scoring_func_int is not None
+    ), f"Unknown scoring_func '{scoring_func}', must be one of {list(_MOE_SCORING_FUNC_MAP.keys())}"
+
```
[the remark this answers is not in the document]
Comment (gaopengff) [reply python/sgl_kernel/moe.py]: Fixed.

File: src/sycl/BiasedTopK.cpp [hunk 2]
```diff
@@ -0,0 +1,524 @@
...
+  TORCH_CHECK(input.size(1) == bias.size(0), "input.size(1) must match bias.size(0)");
+  TORCH_CHECK(
+      input.scalar_type() == torch::kFloat32 || input.scalar_type() == torch::kFloat16 ||
+          input.scalar_type() == torch::kBFloat16,
+      "input must be float32, float16, or bfloat16");
+  TORCH_CHECK(bias.scalar_type() == torch::kFloat32, "bias must be float32");
+  TORCH_CHECK(input.size(1) <= kMaxExperts, "num_experts exceeds maximum supported value: ", kMaxExperts);
+  TORCH_CHECK(topk > num_fused_shared_experts, "topk must be greater than num_fused_shared_experts");
+  TORCH_CHECK(topk <= kMaxTopK, "topk exceeds maximum supported value: ", kMaxTopK);
+  TORCH_CHECK(scoring_func == 0 || scoring_func == 2, "scoring_func must be 0 (sigmoid) or 2 (sqrtsoftplus)");
```
[the remark this answers is not in the document]
Comment (gaopengff) [reply src/sycl/BiasedTopK.cpp]: Current check is enough.

File: src/sycl/BiasedTopK.cpp [hunk 3]
```diff
@@ -0,0 +1,524 @@
...
+};
+
+template <ScoringFunc kScoringFunc>
+static inline float compute_score(float x) {
+  if constexpr (kScoringFunc == ScoringFunc::kSigmoid) {
+    return 1.0f / (1.0f + sycl::native::exp(-x));
+  } else {
+    return sycl::native::sqrt(sycl::native::log(1.0f + sycl::native::exp(x)));
+  }
+}
```
[the remark this answers is not in the document]
Comment (gaopengff) [reply src/sycl/BiasedTopK.cpp:36]: Align to cuda implentation with "sycl::log1p".

File: benchmark/bench_biased_topk.py [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+    if renormalize:
+        topk_weights_sum = (
+            topk_weights.sum(dim=-1, keepdim=True)
+            if num_fused_shared_experts == 0
+            else topk_weights[:, :-1].sum(dim=-1, keepdim=True)
+        )
+        topk_weights = topk_weights / topk_weights_sum
+        if apply_routed_scaling_factor_on_output:
+            topk_weights *= routed_scaling_factor
+
```
[the remark this answers is not in the document]
Comment (gaopengff) [reply benchmark/bench_biased_topk.py:65]: This reference method is from sglang and there is no scenario to use "apply_scale=True, renormalize=False". Keep the current status.

File: tests/test_biased_topk.py [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+    routed_scaling_factor: Optional[float] = None,
+    apply_routed_scaling_factor_on_output: Optional[bool] = False,
+):
+    assert hidden_states.shape[0] == gating_output.shape[0], "Number of tokens mismatch"
+
+    if scoring_func == "sigmoid":
+        scores = gating_output.sigmoid()
+    elif scoring_func == "sqrtsoftplus":
+        scores = torch.nn.functional.softplus(gating_output).sqrt()
+
```
[the remark this answers is not in the document]
Comment (gaopengff) [reply tests/test_biased_topk.py:29]: That's redundant check.

File: src/sycl/BiasedTopK.cpp [hunk 6]
```diff
@@ -0,0 +1,524 @@
...
+enum class ScoringFunc : uint32_t {
+  kSigmoid = 0,
+  kSqrtSoftplus = 2,
+};
+
+template <ScoringFunc kScoringFunc>
+static inline float compute_score(float x) {
+  if constexpr (kScoringFunc == ScoringFunc::kSigmoid) {
+    return 1.0f / (1.0f + sycl::native::exp(-x));
+  } else {
```
Comment (jianan-gu) [inline src/sycl/BiasedTopK.cpp:33]: need to match `kSqrtSoftplus `, not seciently using `else`? and raise unimplemented if necessary

File: python/sgl_kernel/moe.py [hunk 7]
```diff
@@ -143,6 +144,38 @@ def moe_fused_gate(
...
+    indices,
+    topk,
+    scoring_func,
+    num_fused_shared_experts=0,
+    renormalize=False,
+    routed_scaling_factor=1.0,
+    apply_routed_scaling_factor_on_output=False,
+):
+    scoring_func_int = _MOE_SCORING_FUNC_MAP.get(scoring_func.lower())
+    if scoring_func_int is None or scoring_func_int == _MOE_SCORING_FUNC_MAP["softmax"]:
```
Comment (jianan-gu) [inline python/sgl_kernel/moe.py]: why softmax is not included?

File: src/sycl/BiasedTopK.cpp (hunk 6 above)
Comment (gaopengff) [reply src/sycl/BiasedTopK.cpp:33]: Actually the python frontend already done the check, the scoring_func only can be **sigmoid** or **sqrtsoftplus**. https://github.com/sgl-project/sgl-kernel-xpu/blob/pengfei/add_biased_topk/python/sgl_kernel/moe.py#L160

File: python/sgl_kernel/moe.py (hunk 7 above)
Comment (gaopengff) [reply python/sgl_kernel/moe.py]: Cuda does not include softmax either, I guess there is no model using biased_topk with softmax activation. Besides, deepseek-v4 will only use 'sqrtsoftplus' for now.

Comment (jianan-gu) [reply python/sgl_kernel/moe.py]: then we shall not take softmax as a specifical case in codes? here in frontend check and also left` 1 `value for it in kernel

Comment (gaopengff) [reply python/sgl_kernel/moe.py]: It's because frontend end dict _MOE_SCORING_FUNC_MAP is used by others kernels such as grouped_topk, which has softmax support. So the dict contains 'softmax' key. Do you suggest using **"scoring_func == 'sigmoid' or 'sqrtsoftplus'"** to check here?

File: src/sycl/BiasedTopK.cpp [hunk 8]
```diff
@@ -0,0 +1,524 @@
...
+
+    for (uint32_t e = lane_id; e < num_experts_; e += kWarpSize) {
+      const float input_val = static_cast<float>(input_[static_cast<int64_t>(row_idx) * num_experts_ + e]);
+      float bias_val = bias_[e];
+      float score_val = compute_score<kScoringFunc>(input_val);
+      shared_scores[e] = score_val + bias_val;
+      shared_original_scores[e] = score_val;
+    }
+
+    sycl::group_barrier(item.get_group(), sycl::memory_scope::work_group);
```
Comment (jianan-gu) [inline src/sycl/BiasedTopK.cpp]: The work-group barrier seems unnecessary here, please double check.

File: python/sgl_kernel/moe.py (hunk 7 above)
Comment (jianan-gu) [reply python/sgl_kernel/moe.py]: 1. if not in _MOE_SCORING_FUNC_MAP, it is `None `by default?
2. I think just the key here is originally _MOE_SCORING_FUNC_MAP used for `moe_fused_gate` , and is it supporting "sqrtsoftplus"? https://github.com/sgl-project/sgl-kernel-xpu/blob/d0c0b9aec7227207f8dd67697b8b9b3389b186df/python/sgl_kernel/moe.py#L128C5-L128C71

Comment (gaopengff) [reply python/sgl_kernel/moe.py]: Now I added '_MOE_BIASED_TOPK_SCORING_MAP' dict for biased_topk only, which won't affect others.

File: src/sycl/BiasedTopK.cpp (hunk 8 above)
Comment (gaopengff) [reply src/sycl/BiasedTopK.cpp]: Yes, It's redundant here, I've removed it.

**Review by jianan-gu — APPROVED** [review]
